File: python/2025/day02.py
#!/usr/bin/env python3

import re


def part_one(ranges: list[list[int]]) -> int:
    total = 0
    for start, end in ranges:
        for i in range(start, end + 1):
            i_str = str(i)
            i_len = len(i_str)
            if i_str[:i_len//2] == i_str[i_len//2:]:
                total += i

    return total


def part_two(ranges: list[list[int]]) -> int:
    total = 0
    pattern = re.compile(r"^(.+)\1+$")
    for start, end in ranges:
        for i in range(start, end + 1):
            if pattern.match(str(i)):
                total += i

    return total


if __name__ == '__main__':
    with open('../inputs/2025/day02.txt') as file:
        ranges = [list(map(int, x.split('-'))) for x in file.read().rstrip().split(',')]

    print(f"Part 1: {part_one(ranges)}")
    print(f"Part 2: {part_two(ranges)}")

File: cpp/2025/day02.cpp
#include <charconv>
#include <cstddef>
#include <fstream>
#include <iostream>
#include <ranges>
#include <regex>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

long part_one(const std::vector<std::pair<long, long>>& ranges) {
    long total = 0;
    for (const auto& [start, end] : ranges) {
        for (long i = start; i <= end; i++) {
            std::string i_str = std::to_string(i);
            size_t middle = i_str.size() / 2;

            if (i_str.compare(0, middle, i_str, middle, i_str.size()) == 0) {
                total += i;
            }
        }
    }

    return total;
}

long part_two(const std::vector<std::pair<long, long>>& ranges) {
    long total = 0;
    std::regex pattern("^(.+)\\1+$");
    for (const auto& [start, end] : ranges) {
        for (long i = start; i <= end; i++) {
            if (std::regex_search(std::to_string(i), pattern)) {
                total += i;
            }
        }
    }

    return total;
}

int main() {
    std::ifstream file("../inputs/2025/day02.txt");
    std::string data;
    std::getline(file, data);
    std::vector<std::pair<long, long>> input;

    for (const auto pair_view : std::ranges::views::split(data, ',')) {
        std::string_view sv(pair_view.begin(), pair_view.end());
        auto dash_pos = sv.find('-');
        long lhs, rhs;

        std::from_chars(sv.data(), sv.data() + dash_pos, lhs);
        std::from_chars(sv.data() + dash_pos + 1, sv.data() + sv.size(), rhs);
        input.emplace_back(lhs, rhs);
    }

    std::cout << "Part 1: " << part_one(input) << "\nPart 2: " << part_two(input);
}
